prov/rxm/src/rxm_cq.c

```diff
@@ -949,19 +949,31 @@ static inline ssize_t rxm_handle_atomic_resp(struct rxm_ep *rxm_ep,
 
 	tx_buf = ofi_bufpool_get_ibuf(rxm_ep->buf_pools[RXM_BUF_POOL_TX_ATOMIC].pool,
 				      rx_buf->pkt.ctrl_hdr.msg_id);
-	FI_DBG(&rxm_prov, FI_LOG_CQ,
-	       "Received Atomic Response for msg_id: 0x%" PRIx64 "\n",
+	FI_DBG(&rxm_prov, FI_LOG_CQ, "received atomic response: op: %" PRIu8
+	       " msg_id: 0x%" PRIx64 "\n", rx_buf->pkt.hdr.op,
 	       rx_buf->pkt.ctrl_hdr.msg_id);
 
 	assert(!(rx_buf->comp_flags & ~(FI_RECV | FI_REMOTE_CQ_DATA)));
 
-	if (resp_hdr->status) {
-		FI_DBG(&rxm_prov, FI_LOG_CQ,
-		       "Bad Atomic response status %d\n", ntohl(resp_hdr->status));
-		rxm_cq_write_error(rxm_ep->util_ep.tx_cq,
-				   rxm_ep->util_ep.tx_cntr,
+	if (OFI_UNLIKELY(resp_hdr->status)) {
+		struct util_cntr *cntr;
+		FI_WARN(&rxm_prov, FI_LOG_CQ,
+		       "bad atomic response status %d\n", ntohl(resp_hdr->status));
+
+		if (tx_buf->pkt.hdr.op == ofi_op_atomic) {
+			cntr = rxm_ep->util_ep.wr_cntr;
+		} else if (tx_buf->pkt.hdr.op == ofi_op_atomic_compare ||
+			   tx_buf->pkt.hdr.op == ofi_op_atomic_fetch) {
+			cntr = rxm_ep->util_ep.rd_cntr;
+		} else {
+			FI_WARN(&rxm_prov, FI_LOG_CQ,
+				"unknown atomic request op!\n");
+			assert(0);
```

Comment:
Is there a better way to bail here?

Reply:
The assertion would happen only in debug builds. We don't know which counter to report the error on, so we don't do anything here.

Reply:
Since it is an RXM internal implementation error for the op not to be set to one of these 3 values ( which is why the original code did not test values in the else clause), it seems to me an assertion in the else clause that tested for those two values would be sufficient (and the final "else" clause would not be necessary).  i.e.:

if (tx_buf->pkt.hdr.op == ofi_op_atomic) {
 			cntr = rxm_ep->util_ep.wr_cntr;
 } else {
                      assert(tx_buf->pkt.hdr.op == ofi_op_atomic_compare ||
 			   tx_buf->pkt.hdr.op == ofi_op_atomic_fetch);
 			cntr = rxm_ep->util_ep.rd_cntr;
}

Reply:
Although this is an error in rxm, it's coming from a peer.  So, it depends on how much you want to trust that the peer isn't sending invalid or corrupt fields.  Personally, I'm inclined to go with Arun's approach, so that we don't incorrectly update the wrong counter.

Reply:
Unless of course this is an error that's entirely local, then Steve's suggestion is better.  :)

Reply:
@a-ilango This isn't a fatal error in non-debug builds?

Reply:
> This isn't a fatal error in non-debug builds?

This is indeed a fatal error. In the non-debug build, the application would hang. Setting FI_LOG_LEVEL=warn should show the issue. If hang is not desirable, the alternative I see is to modify the endpoint, CQ states to error so that the next operation on the endpoint or the CQ returns a fatal error. We don't maintain such states today but I think that might be helpful.